File: Depricated/Sharp.Data/LinkData.cs
using Dapper;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sharp.Data.Depricated
{
    public class LinkData : ILinkData
    {
        public async Task AddLinkRolePair(ulong guildId, string linkCode, ulong roleId, int uses)
        {
            using var connection = DataExtentions.GetConnection();
            connection.Open();
            await connection.ExecuteAsync($"SELECT add_link_role_pair('{guildId}', '{linkCode}', '{roleId}', {uses})");
        }
        public async Task<List<Link>> GetLinkRolePairsAsync(ulong guildId)
        {
            using var connection = DataExtentions.GetConnection();
            connection.Open();
            var querryResult = await connection.QueryAsync<Link>($"SELECT * FROM link_role_pair WHERE guild_id = '{guildId}'");
            return querryResult.ToList();
        }

        public async Task UpdateUsesAsync(string code, int uses)
        {
            using var connection = DataExtentions.GetConnection();
            connection.Open();
            await connection.ExecuteAsync($"SELECT update_link_uses('{code}', {uses})");
        }
    }
}

File: Sharp.Data/LinkData.fs
namespace Sharp.Data

open Sharp.Domain
open Npgsql.FSharp
open System

module LinkData =

    let private parseLinkRolePair (read : RowReader ) =
        { guildId = read.string "guild_id" |> UInt64.Parse |> GuildId
          roleId = read.string "role_id" |> UInt64.Parse |> RoleId
          linkCode = read.string "code"
          uses = read.int "uses"}

    let addLinkRolePair (GuildId guildId) (RoleId roleId) linkCode uses =
        sprintf "SELECT add_link_role_pair('%i', '%i', '%s', %i)"
            guildId roleId linkCode uses
        |> Operations.executeNonQuery

    let updateUses linkCode uses =
        sprintf "SELECT update_link_uses('%s', %i)"
            linkCode uses
        |> Operations.executeNonQuery

    let getLinkRolePairs (GuildId guildId) =
        sprintf "SELECT * FROM link_role_pair 
                 WHERE guild_id = '%i'"
            guildId
        |> Operations.executeQuery parseLinkRolePair
